add recursive function to find currency dollars

zkamvar opened the pull request:
This adds a stronger solution to #121.

It does a second check to see if there are more opening dollar signs than
closing and then passes those through a function that walks the list of broken
math elements and detects the opening elements that aren't paired with closing
elements (by confirming that the pairs are adjacent). The opening elements that
aren't paired are excluded from the list and then if the list is still not
balanced an error is thrown. 


Update 2025-10-06: I worked on this a bit yesterday and found that I might be able to actually detect the differences between Mal-formed math and stray currency notation in https://github.com/ropensci/tinkr/pull/123/commits/710224353e4c2cdd3ed65134b1f6b1f1d4ae09ad. 

With this PR, we are being greedy with math protection, as we are throwing out math elements that do not conform to our expectations. However, the way we are doing it does not take into account the fact that a node could be marked as ambiguous (has no start and no end markers). I think when we take this into account, we can properly detect broken math elements and throw a warning for them instead of throwing an error.

zkamvar committed:
first pass at stronger solution to #121

This does not work because it breaks the postfix dollar treatment

zkamvar committed:
add example document for testing math parsing

zkamvar committed:
just record length of headless and endless

This makes it easier to re-test once we've transformed

zkamvar committed:
refactor special cases for broken math

1. If there are no headless math, then just return because it's all money
2. If there is are more endless than headless, toss the money and keep
   going

Ultimately, if there is an imbalance (even after 2), then an error
should be thrown

zkamvar committed:
update tests for the currency and postfix dollars

The tests for the postfix dollars have now changed because we are sort
of loosening the restrictions, which ultimately is a good thing because
people will run into fewer errors when protecting math.

zkamvar committed:
refactor recursive trimming function

This simplifies the function to three conditionals from four and clearly
labels the condidtions, starting with the exit condition.

zkamvar committed:
remove unused test file

zkamvar committed:
update documentation

zkamvar committed:
Merge branch 'main' into znk/toss-money/121

maelle commented on R/asis-nodes.R:
```diff
@@ -129,7 +129,7 @@ protect_inline_math <- function(body, ns) {
   }
 
   # protect math that is broken across lines or markdown elements
-  if (length(bmath)) {
```
maybe use an explaining variable although that is already clear

zkamvar replied:
oh good point!

zkamvar replied:
done!

maelle commented on R/asis-nodes.R:
```diff
@@ -140,15 +140,26 @@ protect_inline_math <- function(body, ns) {
     if (lh == 0) {
       return(copy_xml(body))
     }
+    # 2024-10-15: if the number of headless tags is _less_ than the number of
```
```suggestion
    # 2024-10-15: if the number of startless tags is _less_ than the number of
```

maelle replied:
the variable is actually headless, but I wonder if it might be confusing

zkamvar replied:
I think you have a good point and will change the variable name.

maelle commented on R/asis-nodes.R:
```diff
@@ -163,6 +174,58 @@ protect_inline_math <- function(body, ns) {
+  }
+  # return the result and iterate over the rest of the vector
+  return(c(result, toss_broken_teeth(endless[idx], headless[idx])))
+}
```
should we use devtag https://github.com/moodymudskipper/devtag

and have examples

zkamvar replied:
That's a good point. I had avoided it because I though "I don't want to learn a new thing" but then I realized that this will definitely help us in the end (especially if it means that we get local docs that CRAN does not care or know about 😈)

maelle commented on man/rmd-fragments/format-latex.Rmd:
````diff
@@ -19,31 +19,17 @@ Note, however, that there are a few caveats for this:
+      tinkr::yarn$new(ex)$protect_math()$get_protected()
+      ```
+   2. Postfix dollar signs (e.g. using BASIC code demostrations) should be
+      protected with backtics, or it will cause an error.
````
```suggestion
      protected with backticks, or it will cause an error.
```

maelle commented on man/rmd-fragments/format-latex.Rmd:
````diff
@@ -19,31 +19,17 @@ Note, however, that there are a few caveats for this:
+      ex <- textConnection("$5 for _any_ $\\pi$")
+      tinkr::yarn$new(ex)$protect_math()$get_protected()
+      ```
+   2. Postfix dollar signs (e.g. using BASIC code demostrations) should be
````
```suggestion
   2. Postfix dollar signs (e.g. using BASIC code demonstrations) should be
```

zkamvar replied:
I have figured out how to turn on spellcheck in vim, so I can now catch silly things like this! 😅

maelle replied:
nice!!

maelle reviewed (commented):
thank you!! another solution: use € not $ :smile_cat:

zkamvar committed:
Merge branch 'main' into znk/toss-money/121

zkamvar committed:
Merge branch 'main' into znk/toss-money/121

zkamvar committed:
random cleanup

zkamvar committed:
fix failing tests

because of the changes from #124, this error really no longer occurs

I am going to keep it in the codebase and will try to write a test that
will catch this. (I suspect a single prefix followed by three postfix
might do the trick)

zkamvar committed:
Apply suggestions from code review

Co-authored-by: Maëlle Salmon <maelle.salmon@yahoo.se>

zkamvar committed:
change headless to startless

zkamvar committed:
var rename: endles -> no_end; startless -> no_start

zkamvar committed:
add dev doc for toss broken teeth

zkamvar committed:
add note for creating a test

I have been able to get the error to appear with the original math
example by removing one of the dollar signs, but I was unable to
reproduce it using just the header text. I do not have time to debug
this fully right now, so I added a note to myself to come back to this.

zkamvar committed:
re-add test for broken math

zkamvar committed:
add notes about the tests for broken math

zkamvar committed:
fix silly error :see_no_evil:

let this be a lesson to you kids: always clean your workspace before
testing

zkamvar committed:
add xor gate detection of broken elements

This implements the method described in 71022435. We use a XOR
comparison between the no_start and no_end elements to exclude ambiguous
elements. We then use another XOR comparison between that result to
identify the elements that are likely broken math elements and
optionally present a warning.

zkamvar committed:
add todo note to use gregexpr

zkamvar committed:
add failure mode for #146

zkamvar committed:
add note about what is happening with this.
